pkg/events/events.go

```diff
@@ -10,72 +13,209 @@ import (
+//       pointer receivers, and public fields were accessed directly from many places. That happened
+//       because the data never changed and there was no need for concurrency. That is no longer
+//       true, so all the events should now be defined as pointers with private fields and public
+//       getters/setters. This way access to events can be synchronized in the object themselves,
+//       instead of having to do it from the outside (and missing to do so).
+//
+// TODO: Because of dynamically loading/unloading work (and parallelism):
+//
+//       1. Simple types should be atomics (bool, int, uint, pointers, etc.):
+//          - os long as object changes are atomic (no need for multiple field updates)
+//          - usually requires object to have an immutable handle (ID)
+//          - protect loads/stores but not objects single view among 2 CPUs
+//       2. Complex types should have mutexes (maps, slices, etc.)
+//          - updates aren't copy-on-write like atomic types
+//          - complex types are walked through, and not copy-on-write changed
+//       3. All types should be private (no public fields)
+//       4. All types should have getters/setters
+//       5. Return copies of complex types (maps, slices, etc.)
+//
+//
+// Types (keep in a single file for now):
+//
+// 0. ID					immutable
+// 1. EventGroup			mutable, map protected by a mutex
+// 2. Event					mutable with atomic types (no transactional updates needed)
+// 3. Dependencies			mutable with atomic types (no transactional updates needed)
+// 4. ProbeDependency		mutable (atomic "required" field)
+// 5. KSymbolDependency		mutable (atomic "required" field)
+// 6. CapDependency			mutable, map protected by a mutex
+// 7. TailCall				mutable, slice protected by a mutex, atomic strings
+//
+
+// TODO: turn event definitions into EventDefGroup of EventDefs:
+
+//
+// TailCall
+//
+
+const (
+	TailVfsWrite  uint32 = iota // Index of a function to be used in a bpf tailcall.
+	TailVfsWritev               // Matches defined values in ebpf code for prog_array map.
+	TailSendBin
+	TailSendBinTP
+	TailKernelWrite
+	TailSchedProcessExecEventSubmit
+	TailVfsRead
+	TailVfsReadv
+	TailExecBinprm1
+	TailExecBinprm2
+	TailHiddenKernelModuleProc
+	TailHiddenKernelModuleKset
+	TailHiddenKernelModuleModTree
+	TailHiddenKernelModuleNewModOnly
+	MaxTail
+)
+
+type TailCall struct {
+	mapName    *atomic.Pointer[string] // atomic: atomic pointer load/store (get/set)
+	progName   *atomic.Pointer[string] // atomic: atomic pointer load/store (get/set)
+	mapIndexes []uint32                // load/store for array pointer also protected by mutex
+	mutex      *sync.RWMutex           // mutex: protect the array iterations (transactions)
 }
 
-type probeDependency struct {
-	Handle   probes.Handle
-	Required bool // tracee fails if probe can't be attached
+// NewTailCall creates a new TailCall with default values.
+func NewTailCall(mapName string, progName string) *TailCall {
```

Comment:
I don't see any usage of this function except in `NewTailCallFull` below. Any reason not to merge them into `NewTailCall`?
Also, maybe we should introduce a `tailcall` package for type separation and to reduce this file size. WDYT?

Reply:
I'm not sure there's any meaning to initializing a tail call with no indices... If I wanted to construct such a tail call (for some reason) I would pass an empty array. Even then, for the tail call to do anything, it would need to be given some index eventually.
Your call but I think you won't need the "non-full" version.